app/src/main/java/org/lineageos/glimpse/recyclerview/ThumbnailAdapter.kt

```diff
@@ -82,66 +63,54 @@ class ThumbnailAdapter(
         }
 
     override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
-        val truePosition = getTruePosition(position)
-
         when (holder.itemViewType) {
-            ViewTypes.ITEM.ordinal -> {
+            ViewTypes.THUMBNAIL.ordinal -> {
                 val thumbnailViewHolder = holder as ThumbnailViewHolder
-                thumbnailViewHolder.bind(data[truePosition])
+                thumbnailViewHolder.bind((data[position] as DataType.Thumbnail).media)
             }
 
-            ViewTypes.HEADER.ordinal -> {
+            ViewTypes.DATE_HEADER.ordinal -> {
                 val dateHeaderViewHolder = holder as DateHeaderViewHolder
-                dateHeaderViewHolder.bind(data[truePosition].dateAdded)
+                dateHeaderViewHolder.bind((data[position] as DataType.DateHeader).date)
             }
         }
     }
 
-    override fun getItemViewType(position: Int): Int {
-        if (!allowHeaders) {
-            return ViewTypes.ITEM.ordinal
-        }
-
-        if (headersPositions.contains(position)) {
-            return ViewTypes.HEADER.ordinal
-        }
-
-        if (position == 0) {
-            // First element must always be a header
-            addHeaderOffset(position)
-            return ViewTypes.HEADER.ordinal
-        }
-
-        val previousPosition = position - 1
+    override fun getItemViewType(position: Int) = data[position].viewType
 
-        if (headersPositions.contains(previousPosition)) {
-            // Before this position we have a header, next up there's a thumbnail
-            return ViewTypes.ITEM.ordinal
-        }
+    fun setMedias(medias: Array<Media>) {
```

Comment:
Move that logic in the view model.
Doing modeling has no business in the view.

Reply:
Done